ocamltest/main.ml

```diff
@@ -88,69 +88,98 @@ let report_error loc e bt =
   print_exn loc e bt;
   "=> error in test script"
 
-type result_summary = No_failure | Some_failure | All_skipped
-let join_result summary result =
+(* A "path summary" summarizes a list of test results, on a path
+   from the root of the test tree to the current action. *)
+type path_summary =
+  | Some_failure (* one result was a failure *)
+  | All_skipped_so_far (* all were skips *)
+  | Some_pass_no_fail (* at least one Pass, no failure *)
+
+(* A "result summary" summarizes a tree of test results *)
+type result_summary =
+  | Some_success (* One full path had only Pass *)
+  | Some_failure (* One full path had a failure *)
+  | No_success (* No failure, but all paths skipped before the end *)
+
+let join_result (summary : path_summary) result : path_summary =
   let open Result in
   match result.status, summary with
   | Fail, _
   | _, Some_failure -> Some_failure
-  | Skip, All_skipped -> All_skipped
-  | _ -> No_failure
+  | Skip, All_skipped_so_far -> All_skipped_so_far
+  | _ -> Some_pass_no_fail
```

Comment:
My preference would be to remove the wildcard matches and make the matches explicit:
```
let join_result (summary : path_summary) result : path_summary =
  let open Result in
  match result.status, summary with
  | Fail, (Some_failure | All_skipped_so_far | Some_pass_no_fail)
  | (Pass | Skip), Some_failure -> Some_failure
  | Skip, All_skipped_so_far -> All_skipped_so_far
  | (Pass | Skip), (All_skipped_so_far | Some_pass_no_fail) ->
     Some_pass_no_fail
```

Reply:
I changed the code to simplify the summary a bit. At the same time I went for a compromise between your very-explicit proposal and my very-implicit version:

```ocaml
let join_result (summary : path_summary) result : path_summary =
  let open Result in
  match result.status, summary with
  | Fail, _ | _, Some_failure -> Some_failure
  | Pass, All_pass -> All_pass
  | Pass, Some_skip
  | Skip, (All_pass | Some_skip) -> Some_skip

let join_summaries sa sb =
  match sa, sb with
  | Some_failure, _ | _, Some_failure -> Some_failure
  | Some_success, _ | _, Some_success -> Some_success
  | No_success, No_success -> No_success
```

In `join_result`, I find it clear that failure is an absorbing element, that is, that failure joined with anything is failure. On the other hand, it is less obvious which cases result in "Some skip", so I expanded in details as you suggested. In `join_summary`, failure is absorbing as before, and then success is absorbing with the remaining, non-failing summaries (success or not, but no failure).